### DISCUSSION TITLE
How to mock a module multiple times?

### DISCUSSION THREAD

**Discussion by tal-rofe** [vitest-dev/vitest#1255] [Q&A] [answered]
In my source code, I use the `cosmiconfig` module. My code integrates with it, and I'd like to test some cases with it.

Here is my source code:
```ts
import { cosmiconfig } from 'cosmiconfig';
import TypeScriptLoader from 'cosmiconfig-typescript-loader';

const explorer = cosmiconfig(CONFIGURATION_MODULE_NAME, {
	searchPlaces: SEARCH_PLACES,
	loaders: {
		'.ts': TypeScriptLoader(),
	},
});

const result = await explorer.search();
```

I want to mock this `cosmiconfig` module in order to test some cases.

What's the correct way of doing so?

For example, for checking upon `null` return value from `await explorer.search()`, I tried:
```ts
import { describe, it, expect, afterEach, vi } from 'vitest';

vi.mock('cosmicocnfig', () => {
	return {
		cosmiconfig: vi.fn(() => {
			return {
				search: () => Promise.resolve(null),
			};
		}),
	};
});
```

But now I also want to test when `await explorer.search()` rejects. How can I mock it again?

**Comment by sheremet-va** [comment, marked as answer]
You can import it and change the returned value:

```ts
import { cosmiconfig } from 'cosmiconfig'

vi.mocked(cosmiconfig).mockReturnValue({
  search: () => Promise.reject(null)
})
```

`vi.mocked` is just a typehint, btw. It's noop.

**Reply by tal-rofe** [reply]
Thanks. Should I run `vi.restoreAllMocks()` in `afterEach` hook?

By the way I ran it as standalone block (without my code, using `vi.mock`) but I got an error:
`TypeError: __vite_ssr_import_0__.vi.mocked(...).mockReturnValue is not a function`

Here is what I've tried:
https://gist.github.com/tal-rofe/424e77e1d2caa2aa9315e7bb6493b20a

**Reply by sheremet-va** [reply]
I don't see you calling `vi.mock` to mock the module itself?

**Reply by tal-rofe** [reply]
@sheremet-va 
Yes. I mean, I didn't do so. Should I?

**Reply by sheremet-va** [reply]
Well, how else should Vitest know it's mocked? 
As I said, `vi.mocked` is just a typehint. It doesn't do anything. Just provides types.

**Reply by tal-rofe** [reply]
@sheremet-va 

So, will the following snippet do the job?

```ts
import { describe, it, expect, afterEach, vi } from 'vitest';

vi.mock('cosmicocnfig');
it('getConfiguration | Function should return the default when there is no configuration file', async () => {
vi.mocked(cosmiconfig).mockReturnValue({
	search: () => Promise.resolve(null),
	...cosmiconfigFunctions,
});

const result = await getConfiguration();

expect(result).toJsonEqual(DEFAULT_CONFIGURATION);
});

```ts

**Reply by sheremet-va** [reply]
I think, yes.

**Reply by tal-rofe** [reply]
@sheremet-va 

I still get an error with the following snippet:
https://gist.github.com/tal-rofe/26ec299a1b266d49d880e690781450e4

`TypeError: __vite_ssr_import_0__.vi.mocked(...).mockReturnValue is not a function`

**Reply by sheremet-va** [reply]
What gives this error? I see you are doing these:
```
vi.mocked(validateConfiguration).mockReturnValue({});
```

but you haven't mocked the module you are importing them from

**Reply by tal-rofe** [reply]
For example,

```
TypeError: __vite_ssr_import_0__.vi.mocked(...).mockReturnValue is not a function
 ❯ tests/utils/configuration.spec.ts:144:26
    142|                        };
    143| 
    144|                        vi.mocked(cosmiconfig).mockReturnValue({
       |                          ^
    145|                                search: () =>
    146|                                        Promise.resolve({
```

**Comment by tal-rofe** [comment]
@sheremet-va 

I know I have a flaw in line 8 in the snippet I've provided. I fixed the flaw but the issue persists

**Comment by tal-rofe** [comment]
Can confirm the answer was indeed provided, and I have some irrelevant issue

**Comment by jrh-crafted** [comment]
The following works for me.

**myFunction.js**

```
import library from '@whatever/library'

export default () => {
  return library()
}
```

**myFunction.spec.js**

```
import { describe, expect, it, vi } from 'vitest'

// --------------------------------------------

import library from '@whatever/library'
vi.mock('@whatever/library')

// This also seems to work in my environment, but it sounds like this may not be officially supported or
// work in all environments.
// 
// const library = vi.fn()
// vi.mock('@whatever/library', () => ({ default: library }))

// --------------------------------------------

import myFunction from './myFunction.js'

// --------------------------------------------

describe('myFunction', () => {
  it('returns value from library', () => {
    library.mockReturnValue('foo')
    expect(myFunction()).toEqual('foo')

    library.mockReturnValue('bar')
    expect(myFunction()).toEqual('bar')
  })
})
```

**Reply by sheremet-va** [reply]
> // const library = vi.fn()
> // vi.mock('@whatever/library', () => ({ default: library }))

This won't work, because `vi.mock` is hoisted, and defined before `library`.

**Reply by jrh-crafted** [reply]
Perhaps it shouldn't, but it does – at least in my environment. 🙃 I've just tested it.

`vi.mock` throws the expected `ReferenceError` if the `library` variable is used within the call to `vi.mock`, i.e.

```
const library = vi.fn()
vi.mock('svelte', library)
```

However, if the `library` variable is referenced inside the anonymous function passed to `vi.mock` as above, it works just fine for me.

Regardless, it sounds like the best practice would be the pattern suggested by @sheremet-va. 👍🏼 Updated my answer to include this caveat.

**Comment by RossRogers** [comment]
See also https://github.com/vitest-dev/vitest/discussions/4124
